File: src/main/java/com/jesperancinha/atm/finder/service/ATMLocatorServiceImpl.java
package com.jesperancinha.atm.finder.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.jesperancinha.atm.finder.service.payload.response.ATMMachine;
import lombok.Setter;
import org.apache.camel.BeanInject;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.RestTemplate;

import java.io.IOException;
import java.util.Arrays;

/**
 * Created by joaofilipesabinoesperancinha on 28-07-16.
 */

@Setter
public class ATMLocatorServiceImpl implements ATMLocatorService {

    @BeanInject
    private RestTemplate restTemplate;

    @BeanInject
    private ObjectMapper mapper;

    private String atmEndpoint;

    public ATMMachine[] getAtmPerCity(String city) throws IOException {
        ResponseEntity<String> response = restTemplate
                .getForEntity(
                        atmEndpoint,
                        String.class);

        String body = response.getBody();
        body = body.substring(body.indexOf('\n') + 1);

        ATMMachine[] results = mapper.readValue(body, ATMMachine[].class);

        return Arrays
                .stream(results)
                .filter(atmMachine -> atmMachine.getAddress().getCity().equals(city))
                .toArray(ATMMachine[]::new);
    }
}

File: jeorg-camel-app-atm/src/main/kotlin/org/jesperancinha/atm/finder/service/AtmLocatorServiceImpl.kt
package org.jesperancinha.atm.finder.service

import com.fasterxml.jackson.databind.ObjectMapper
import lombok.Setter
import org.apache.camel.BeanInject
import org.jesperancinha.atm.finder.service.payload.response.ATMMachine
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import org.springframework.web.client.RestTemplate
import java.io.IOException
import java.util.*

/**
 * Created by joaofilipesabinoesperancinha on 28-07-16.
 */
@Setter
@Service(value = "atmLocatorService")
class AtmLocatorServiceImpl {
    @BeanInject
    lateinit var restTemplate: RestTemplate

    @BeanInject
    lateinit var  mapper: ObjectMapper

    @Value("\${atm.endpoint}")
    lateinit var  atmEndpoint: String

    @Throws(IOException::class)
    fun getAtmPerCity(city: String): Array<ATMMachine> {
        val response = restTemplate.getForEntity(
            atmEndpoint,
            String::class.java
        )
        var body = requireNotNull(response.body)
        if (!body.startsWith("{") && !body.startsWith("[")) {
            body = body.substring(body.indexOf('\n') + 1)
        }
        val results = mapper.readValue(body, Array<ATMMachine>::class.java)
        return results
            .filter { atmMachine: ATMMachine -> atmMachine.address.city == city }
            .toTypedArray()
    }
}